packages/flutter_test/lib/src/binding.dart

```diff
@@ -1095,28 +1095,45 @@ class AutomatedTestWidgetsFlutterBinding extends TestWidgetsFlutterBinding {
 
     addTime(additionalTime);
 
-    return realAsyncZone.run<Future<T?>>(() async {
+    return realAsyncZone.run<Future<T?>>(() {
+      final Completer<T?> result = Completer<T?>();
       _pendingAsyncTasks = Completer<void>();
-      T? result;
       try {
-        result = await callback();
+        callback().then((T? value) {
+          result.complete(value);
+        }).catchError((Object exception, StackTrace stack) {
+          FlutterError.reportError(FlutterErrorDetails(
+            exception: exception,
+            stack: stack,
+            library: 'Flutter test framework',
+            context: ErrorDescription('while running async test code'),
+            informationCollector: () {
+              return <DiagnosticsNode>[
+                ErrorHint('The exception was caught asynchronously.'),
+              ];
+            },
+          ));
+          result.complete(null);
+        });
       } catch (exception, stack) {
         FlutterError.reportError(FlutterErrorDetails(
           exception: exception,
           stack: stack,
           library: 'Flutter test framework',
           context: ErrorDescription('while running async test code'),
+            informationCollector: () {
+              return <DiagnosticsNode>[
+                ErrorHint('The exception was caught synchronously.'),
+              ];
+            },
         ));
-      } finally {
-        // We complete the _pendingAsyncTasks future successfully regardless of
-        // whether an exception occurred because in the case of an exception,
-        // we already reported the exception to FlutterError. Moreover,
-        // completing the future with an error would trigger an unhandled
-        // exception due to zone error boundaries.
+        result.complete(null);
+      }
+      result.future.whenComplete(() {
         _pendingAsyncTasks!.complete();
         _pendingAsyncTasks = null;
-      }
-      return result;
+      });
+      return result.future;
```

Comment:
cc @lrhn any idea why this fixes the bug? the bug being https://github.com/flutter/flutter/pull/107874/files#r925041544 hanging in the old version.

Reply:
Probably error zones.
The `realAsyncZone` runes code in the root zone. That's a no-no when running tests, since tests run in a test zone which captures errors, so you are dodging the test framework.
That also means that the future created inside the `realAsyncZone` is in the root zone, a different error zone from the surrounding code.
With this change, the completer is created outside the `realAsyncZone`-run code, so it belongs in the current zone.
And the test below throws.

Reply:
yeah, dodging the test framework is very much the point of this function. The test framework does FakeAsync work and this function is used to call real async functions.

I don't understand why this change changes where the completer is created though. It seems like in both cases it's created in the same zone (i.e. whatever zone is calling this method).